### PR TITLE
Authentication

### PR DESCRIPTION
**Pull request by Ollynov**
SIGNUP:
render on page √
check for pre-existing users, confirmed password √
hash-password √
save to database √

SIGNIN:
render sign in √
authenticate user and keep them signed in √
create new component showing current status('you are signed in as: username')
create new component for sign out button

### CODE REVIEW

File: client/Register.jsx [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    this.handleUsernameInput = this.handleUsernameInput.bind(this);
+    this.handlePasswordInput = this.handlePasswordInput.bind(this);
+    this.handleConfirmPasswordInput = this.handleConfirmPasswordInput.bind(this);
+    this.handleSuccess = this.handleSuccess.bind(this);
+    this.passwordTest = this.passwordTest.bind(this);
+  }
+
+  addUser() {
+    $.ajax({
+      url: 'http://localhost:3000/users',
```
Comment (andrewk17) [inline client/Register.jsx:27]: this should be a relative path. hardcoding to localhost:3000 will break deployment.

please change to just /users

File: client/Register.jsx [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+      type: 'POST',
+      data: JSON.stringify({ currentUsername: this.state.currentUsername,
+        currentPassword: this.state.currentPassword }),
+      contentType: 'application/json; charset=utf-8',
+      success: function(data) {
+        this.handleSuccess();
+      }.bind(this)
```
Comment (andrewk17) [inline client/Register.jsx:34]: i'm wondering if we use arrow function, we may not need this bind

File: client/Register.jsx [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    }).fail(function(result) {
+      this.rejectUserTakenUsername();
+    }.bind(this));
```
Comment (andrewk17) [inline client/Register.jsx:37]: same here with arrow function

File: client/Signin.jsx [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+  handleSubmit(e) {
+    e.preventDefault();
+    $.ajax({
+      url: 'http://localhost:3000/login',
+      type: 'POST',
+      data: JSON.stringify({ username: this.state.currentUsername, password: this.state.currentPassword }),
+      contentType: 'application/json; charset=utf-8',
+      success: function(data) {
+        this.handleSuccess();
+      }.bind(this)
```
Comment (andrewk17) [inline client/Signin.jsx:43]: try arrow function

File: client/Signin.jsx [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    }).fail(function(result) {
+      // 305 error corresponds to wrong password
+      // 300 error corresponds to no such username found
+      if (result.status === 305) {
+        this.rejectSignInWrongPassword();
+      }
+      if (result.status === 300) {
+        this.rejectSignInWrongUsername();
+      }
+    }.bind(this));
```
Comment (andrewk17) [inline client/Signin.jsx:53]: try arrow function

File: client/Register.jsx [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    this.setState({currentPassword: e.target.value});
+  }
+  handleConfirmPasswordInput(e) {
+    this.setState({currentConfirmation: e.target.value});
+  }
+
+  render() {
+    return (
+      <div>
+        <form onSubmit={this.handleSubmit.bind(this)}>
```
Comment (andrewk17) [inline client/Register.jsx:87]: isn't the this binding handled in constructor?

File: client/Signin.jsx [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    this.setState({currentUsername: e.target.value});
+  }
+  handlePasswordInput(e) {
+    this.setState({currentPassword: e.target.value});
+  }
+
+  render() {
+    return (
+      <div>
+        <form onSubmit={this.handleSubmit.bind(this)}>
```
Comment (andrewk17) [inline client/Signin.jsx:66]: is this binding handled in constructor?

File: client/index.jsx [hunk 8]
```diff
@@ -22,13 +26,37 @@ class App extends React.Component {
...
+
   render () {
-    return (
+    if (this.state.register) {
+      return (
       <div>
         <SearchBar setMarkers={this.setMarkers}/>
+        <h1>Register:</h1>
+        <Register /><br/><br/>
+        <a onClick={this.handleSignIn.bind(this)} href='#'>Already have an account? Sign in here</a>
```
Comment (andrewk17) [inline client/index.jsx:44]: should we bind this in constructor to be consistent across files?

File: client/index.jsx [hunk 9]
```diff
@@ -22,13 +26,37 @@ class App extends React.Component {
...
         <div id='mapContainer'><MapComponent markers={this.state.markers} /></div>
       </div>
     );
+    } else {
+      return (
+      <div>
+        <SearchBar setMarkers={this.setMarkers}/>
+        <h1>Sign in:</h1>
+        <Signin />
+        <a onClick={this.handleRegister.bind(this)} href='#'>Register</a>
```
Comment (andrewk17) [inline client/index.jsx:54]: should we bind this in constructor to be consistent across files?

File: server/database/users/userdbcontroller.js [hunk 10]
```diff
@@ -0,0 +1,68 @@
+'use strict';
+
+let User = require('./userschema.js');
+var bcrypt = require('bcryptjs');
```
Comment (andrewk17) [inline server/database/users/userdbcontroller.js:4]: let instead of var

**Comment by andrewk17** [comment]
Nice work. I like it. Can we refactor to ES6 (i.e. arrow functions, let insted of var). Otherwise looks good

File: client/Register.jsx (hunk 6 above)
Comment (Ollynov) [reply client/Register.jsx:87]: Nice catch, I removed it.

**Pull request merged**